File: INCLService/SPCUtility.pas
unit SPCUtility;

interface

uses Classes, sysUtils;

type
  TSPCSchuss = class;
  TSPCValue = class;
  TSPCSchussList = class;
  TSPCValueList = class;

  TSPCMaschine = class
  public
    MaschNr : Integer;
    Lizenz : string;

    SchussList : TSPCSchussList;

    function IsOKFirst : Boolean;
    function IsOKLast : Boolean;
    function getErrorList : TStringList;
    procedure Clear;

    constructor create;
    destructor Destroy;
  end;

  TSPCSchuss = class
    Nr : Integer;
    Schuss : Integer;
    ValueList : TSPCValueList;

    function IsOK : Boolean;

    constructor create;
    destructor Destroy;
  end;

  TSPCSchussList = class(TList)
  private
    function getItem(index: Integer): TSPCSchuss;
    procedure setItem(index: Integer; const Value: TSPCSchuss);

  public
    property Items[index: Integer]: TSPCSchuss read getItem write setItem;

    function Add(aSchuss: TSPCSchuss): Integer;
    function getByNr(aNr : integer) : TSPCSchuss;

    procedure ClearList;
    constructor Create;
    destructor Destroy; override;
  end;

  TSPCValue = class
  private
    FNr : Integer;
    FName : string;
    FSoll : Extended;
    FIst : Extended;
    FMax : Extended;
    FMin : Extended;

    function getMinInPct : Extended;
    procedure setMinInPct(Value : Extended);
    function getMaxInPct : Extended;
    procedure setMaxInPct(Value : Extended);

  public
    property Nr : Integer read FNr write FNr;
    property Name : string read FName write FName;
    property Soll : Extended read FSoll write FSoll;
    property Ist : Extended read FIst write FIst;
    property MaxInPct : Extended read getMaxInPct write setMaxInPct;
    property MinInPct : Extended read getMinInPct write setMinInPct;

    property MaxValue : Extended read FMax write FMax;
    property MinValue : Extended read FMin write FMin;

    function OK : Boolean;
  end;

  TSPCValueList = class(TList)
  private
    function getItem(index: Integer): TSPCValue;
    procedure setItem(index: Integer; const Value: TSPCValue);

  public
    property Items[index: Integer]: TSPCValue read getItem write setItem;

    function Add(aValue: TSPCValue): Integer;
    function getByNr(aNr : integer) : TSPCValue;

    procedure ClearList;
    constructor Create;
    destructor Destroy; override;

  end;


implementation

constructor TSPCMaschine.create;
begin
  SchussList := TSPCSchussList.Create;
end;

destructor TSPCMaschine.Destroy;
begin
  SchussList.Destroy;
  inherited;
end;


function TSPCMaschine.IsOKFirst : Boolean;
begin
  Result := SchussList.Items[0].IsOK;
end;

function TSPCMaschine.IsOKLast : Boolean;
begin
  Result := SchussList.Items[SchussList.Count-1].IsOK;
end;

function TSPCMaschine.getErrorList : TStringList;
var i : Integer;
begin
  result := TStringList.Create;
  for i := 0 to SchussList.Count-1 do
    if not SchussList.Items[i].IsOK then
      Result.Add(IntToStr(SchussList.Items[i].Nr));
  if result.Count = 0 then
  begin
    Result.Destroy;
    Result := nil;
  end;
end;

procedure TSPCMaschine.Clear;
begin
  Lizenz := '';
  MaschNr := 0;
  SchussList.ClearList;
end;

function TSPCSchuss.IsOK : Boolean;
var i : Integer;
begin
  result := True;
  for i := 0 to ValueList.Count -1 do
    result := result and ValueList.Items[i].OK;
end;

constructor TSPCSchuss.create;
begin
  inherited;
  ValueList := TSPCValueList.Create;
end;

destructor TSPCSchuss.Destroy;
begin
  ValueList.Destroy;
  inherited;
end;

function TSPCSchussList.getItem(index: Integer): TSPCSchuss;
begin
  Result := inherited Items[index];
end;

procedure TSPCSchussList.setItem(index: Integer; const Value: TSPCSchuss);
begin
  inherited Items[index]:= Value;
end;

function TSPCSchussList.Add(aSchuss: TSPCSchuss): Integer;
begin
  result := inherited Add(aSchuss);
end;

function TSPCSchussList.getByNr(aNr : integer) : TSPCSchuss;
var i : Integer;
begin
  result := nil;
  for i:= 0 to Count-1 do
    if Items[i].Nr = aNr then
      Result := Items[i];
end;

procedure TSPCSchussList.ClearList;
var i : Integer;
begin
  for i := 0 to Count -1 do
    Items[i].ValueList.ClearList;
  inherited Clear;
end;

constructor TSPCSchussList.Create;
begin
  inherited;
end;

destructor TSPCSchussList.Destroy;
var i : Integer;
begin
  for i := 0 to Count -1 do
  begin
    if Items[i] <> nil then
    begin
      if Items[i] is TSPCSchuss then
        (Items[i] as TSPCSchuss).Destroy;
      Items[i] := nil;
    end;
  end;

  inherited;
end;


function TSPCValue.getMinInPct : Extended;
begin
  result := (1 - (FMin / FSoll)) * 100;
end;

procedure TSPCValue.setMinInPct(Value : Extended);
begin
  FMin := FSoll *(1- (Value / 100));
end;

function TSPCValue.getMaxInPct : Extended;
begin
  result := ((FMax / FSoll) - 1) * 100;
end;

procedure TSPCValue.setMaxInPct(Value : Extended);
begin
  FMax := FSoll *(1+ (Value / 100));

end;


function TSPCValue.OK : Boolean;
begin
  result := (FIst > FMin) and (FIst <= FMax);
end;


function TSPCValueList.getItem(index: Integer): TSPCValue;
begin
  Result := TSPCValue(inherited Items[index]);
end;

procedure TSPCValueList.setItem(index: Integer; const Value: TSPCValue);
begin
  inherited Items[index]:= Value;
end;

function TSPCValueList.Add(aValue: TSPCValue): Integer;
begin
  result := inherited Add(aValue);
end;

function TSPCValueList.getByNr(aNr : integer) : TSPCValue;
var i : Integer;
begin
  result := nil;
  for i:= 0 to Count-1 do
    if Items[i].FNr = aNr then
      Result := Items[i];
end;

procedure TSPCValueList.ClearList;
begin
  inherited Clear;
end;

constructor TSPCValueList.Create;
begin
  inherited;
end;

destructor TSPCValueList.Destroy;
var i : Integer;
begin
  for i := 0 to Count -1 do
  begin
    if Items[i] <> nil then
    begin
      if Items[i] is TSPCValue then
        (Items[i] as TSPCValue).Destroy;
      Items[i] := nil;
    end;
  end;
  inherited;
end;



end.

File: INCLService_CSharp/SPCUtility.cs
using System;
using System.Collections.Generic;

namespace INCLService_CSharp
{
    public class TSPCMaschine : IDisposable
    {
        public int MaschNr { get; set; } = 0;
        public string Lizenz { get; set; } = "";
        public TSPCSchussList SchussList { get; set; }

        public TSPCMaschine()
        {
            SchussList = new TSPCSchussList();
        }

        public bool IsOKFirst()
        {
            if (SchussList.Count > 0)
                return SchussList[0].IsOK;
            return false;
        }

        public bool IsOKLast()
        {
            if (SchussList.Count > 0)
                return SchussList[SchussList.Count - 1].IsOK;
            return false;
        }

        public List<string> GetErrorList()
        {
            List<string> result = new List<string>();
            for (int i = 0; i < SchussList.Count; i++)
            {
                if (!SchussList[i].IsOK)
                {
                    result.Add(SchussList[i].Nr.ToString());
                }
            }
            return result.Count > 0 ? result : null;
        }

        public void Clear()
        {
            Lizenz = "";
            MaschNr = 0;
            SchussList.ClearList();
        }

        public void Dispose()
        {
            if (SchussList != null)
            {
                SchussList.Dispose();
                SchussList = null;
            }
        }

        ~TSPCMaschine()
        {
            Dispose();
        }
    }

    public class TSPCSchuss : IDisposable
    {
        public int Nr { get; set; } = 0;
        public int Schuss { get; set; } = 0;
        public TSPCValueList ValueList { get; set; }

        public TSPCSchuss()
        {
            ValueList = new TSPCValueList();
        }

        public bool IsOK()
        {
            bool result = true;
            for (int i = 0; i < ValueList.Count; i++)
            {
                result = result && ValueList[i].OK();
            }
            return result;
        }

        public void Dispose()
        {
            if (ValueList != null)
            {
                ValueList.Dispose();
                ValueList = null;
            }
        }

        ~TSPCSchuss()
        {
            Dispose();
        }
    }

    public class TSPCSchussList : List<TSPCSchuss>, IDisposable
    {
        public new TSPCSchuss this[int index]
        {
            get => base[index];
            set => base[index] = value;
        }

        public int Add(TSPCSchuss aSchuss)
        {
            base.Add(aSchuss);
            return base.Count - 1;
        }

        public TSPCSchuss GetByNr(int aNr)
        {
            for (int i = 0; i < base.Count; i++)
            {
                if (base[i].Nr == aNr)
                    return base[i];
            }
            return null;
        }

        public void ClearList()
        {
            for (int i = 0; i < base.Count; i++)
            {
                base[i].ValueList.ClearList();
            }
            base.Clear();
        }

        public void Dispose()
        {
            ClearList();
        }

        ~TSPCSchussList()
        {
            Dispose();
        }
    }

    public class TSPCValue : IDisposable
    {
        private int FNr = 0;
        private string FName = "";
        private double FSoll = 0.0;
        private double FIst = 0.0;
        private double FMax = 0.0;
        private double FMin = 0.0;

        public int Nr { get => FNr; set => FNr = value; }
        public string Name { get => FName; set => FName = value; }
        public double Soll { get => FSoll; set => FSoll = value; }
        public double Ist { get => FIst; set => FIst = value; }

        public double MaxInPct
        {
            get => ((FMax / FSoll) - 1) * 100;
            set => FMax = FSoll * (1 + (value / 100));
        }

        public double MinInPct
        {
            get => (1 - (FMin / FSoll)) * 100;
            set => FMin = FSoll * (1 - (value / 100));
        }

        public double MaxValue { get => FMax; set => FMax = value; }
        public double MinValue { get => FMin; set => FMin = value; }

        public bool OK()
        {
            return (FIst > FMin) && (FIst <= FMax);
        }

        public void Dispose()
        {
            // Nothing to dispose
        }

        ~TSPCValue()
        {
            Dispose();
        }
    }

    public class TSPCValueList : List<TSPCValue>, IDisposable
    {
        public new TSPCValue this[int index]
        {
            get => base[index];
            set => base[index] = value;
        }

        public int Add(TSPCValue aValue)
        {
            base.Add(aValue);
            return base.Count - 1;
        }

        public TSPCValue GetByNr(int aNr)
        {
            for (int i = 0; i < base.Count; i++)
            {
                if (base[i].Nr == aNr)
                    return base[i];
            }
            return null;
        }

        public void ClearList()
        {
            base.Clear();
        }

        public void Dispose()
        {
            ClearList();
        }

        ~TSPCValueList()
        {
            Dispose();
        }
    }
}
